django/core/handlers/asgi.py

```diff
@@ -179,15 +180,51 @@ async def handle(self, scope, receive, send):
             body_file.close()
             await self.send_response(error_response, send)
             return
-        # Get the response, using the async mode of BaseHandler.
+        # Try to catch a disconnect while getting response
+        try:
+            response = await self.disconnect_and_get_response(request, receive)
+        except RequestAborted:
+            body_file.close()
+            return
+        # Send the response.
+        await self.send_response(response, send)
+
+    async def listen_for_disconnect(self, receive):
+        """Listens for disconnect from the client."""
+        while True:
+            message = await receive()
+            if message["type"] == "http.disconnect":
+                raise RequestAborted()
+
+    async def disconnect_and_get_response(self, request, receive):
+        """Checks for client disconnect while getting response."""
+        tasks = [
+            asyncio.ensure_future(coro(arg))
+            for coro, arg in zip(
+                (
+                    self.run_get_response,
+                    self.listen_for_disconnect,
+                ),
+                (request, receive),
+            )
+        ]
```

Comment:
This structure is too complex. The `zip` is overkill. 
Why `ensure_future` rather than `create_task`? 

```
tasks = [
    asyncio.create_task(self.listen_for_disconnect(receive)),
    asyncio.create_task(self.run_get_response(request)),
]
```